waku/waku_archive/archive.nim

```diff
@@ -64,186 +75,148 @@ method validate*(validator: DefaultMessageValidator, msg: WakuMessage): Validati
 
-# Retention policy
-const WakuArchiveDefaultRetentionPolicyInterval* = chronos.minutes(30)
+proc periodicRetentionPolicy(self: WakuArchive) {.async.} =
+  debug "executing message retention policy"
 
-proc loopApplyRetentionPolicy*(w: WakuArchive):
-                               Future[Result[void, string]] {.async.} =
+  let policy = self.retentionPolicy.get()
 
-  if w.retentionPolicy.isNil():
-    return err("retentionPolicy is Nil in executeMessageRetentionPolicy")
+  while true:
+    await sleepAsync(WakuArchiveDefaultRetentionPolicyInterval)
 
-  if w.driver.isNil():
-    return err("driver is Nil in executeMessageRetentionPolicy")
+    (await policy.execute(self.driver)).isOkOr:
+      waku_archive_errors.inc(labelValues = [retPolicyFailure])
+      error "failed execution of retention policy", error=error
 
+proc periodicMetricReport(self: WakuArchive) {.async.} =
   while true:
-    debug "executing message retention policy"
-    let retPolicyRes = await w.retentionPolicy.execute(w.driver)
-    if retPolicyRes.isErr():
-        waku_archive_errors.inc(labelValues = [retPolicyFailure])
-        error "failed execution of retention policy", error=retPolicyRes.error
-
-    await sleepAsync(WakuArchiveDefaultRetentionPolicyInterval)
+    await sleepAsync(WakuArchiveDefaultMetricsReportInterval)
 
-  return ok()
+    let count = (await self.driver.getMessagesCount()).valueOr:
+      error "loopReportStoredMessagesMetric failed to get messages count", error=error
+      continue
 
-# Metrics reporting
-const WakuArchiveDefaultMetricsReportInterval* = chronos.minutes(1)
+    waku_archive_messages.set(count, labelValues = ["stored"])
 
-proc loopReportStoredMessagesMetric*(w: WakuArchive):
-                                     Future[Result[void, string]] {.async.} =
-  if w.driver.isNil():
-    return err("driver is Nil in loopReportStoredMessagesMetric")
+proc start*(self: WakuArchive) =
+  if self.retentionPolicy.isSome():
+    self.retentionPolicyHandle = periodicRetentionPolicy(self)
 
-  while true:
-    let resCount = await w.driver.getMessagesCount()
-    if resCount.isErr():
-      return err("loopReportStoredMessagesMetric failed to get messages count: " & resCount.error)
+  self.metricsHandle = periodicMetricReport(self)
 
-    waku_archive_messages.set(resCount.value, labelValues = ["stored"])
-    await sleepAsync(WakuArchiveDefaultMetricsReportInterval)
+proc stopWait*(self: WakuArchive) {.async.} =
+  var futures: seq[Future[void]]
 
-  return ok()
+  if self.retentionPolicy.isSome() and not self.retentionPolicyHandle.isNil():
+    futures.add(self.retentionPolicyHandle.cancelAndWait())
 
-proc start*(self: WakuArchive) {.async.} =
-  ## TODO: better control the Result in case of error. Now it is ignored
-  self.retPolicyFut = self.loopApplyRetentionPolicy()
-  self.retMetricsRepFut = self.loopReportStoredMessagesMetric()
+  if not self.metricsHandle.isNil:
+    futures.add(self.metricsHandle.cancelAndWait())
```

Comment:
Just to learn, what is the benefit of following that approach?

Reply:
Seams like a good way to graceful shutdown when there are many futures to cancel.

- concurrent cancel on all futures
- await but prevent cancellation of the shutdown process

I found it at https://github.com/status-im/nim-eth/blob/d66a29db7ca4372dba116928f979e92cb7f7661f/eth/p2p/discoveryv5/protocol.nim#L1039